charmcraft/services/charmlibs.py

```diff
@@ -0,0 +1,95 @@
+#
+# For further info, check https://github.com/canonical/charmcraft
+"""Service class for interacting with charm libraries."""
+from __future__ import annotations
+
+import pathlib
+
+import craft_application
+
+from charmcraft import models, utils
+from charmcraft.store.models import Library
+
+
+class CharmLibsService(craft_application.ProjectService):
+    """Business logic for creating packages."""
+
+    _project: models.CharmcraftProject  # type: ignore[assignment]
+
+    def __init__(
+        self,
+        app: craft_application.AppMetadata,
+        services: craft_application.ServiceFactory,
+        *,
+        project: models.CharmcraftProject,
+        project_dir: pathlib.Path,
+    ) -> None:
+        super().__init__(app, services, project=project)
+        self._project_dir = project_dir
+
+    def is_lib_downloaded(
+        self, *, charm_name: str, lib_name: str, api: int, patch: int | None = None
+    ) -> bool:
+        """Check if the given charm lib is already downloaded on disk.
+
+        :param charm_name: The name of the charm the lib is attached to.
+        :param lib_name: The name of the lib itself.
+        :param api: The api version of the lib
+        :param patch: If given, the specific patch version of the lib.
+        """
+        lib_path = utils.get_lib_path(charm_name, lib_name, api)
+        if not (self._project_dir / lib_path).exists():
+            return False
+
+        if patch is None:
+            return True
+
+        lib_info = utils.get_lib_info(lib_path=self._project_dir / lib_path)
+        return lib_info.patch == patch
+
+    def get_lib_version(self, *, charm_name: str, lib_name: str) -> tuple[int, int] | None:
+        """Get the version of the library on the machine, or None.
+
+        :param charm_name: The name of the charm where the lib is published
+        :param lib_name: The name of the library itself
+        :returns: Either the version of the library as a pair of integers or None
+            if the library cannot be found.
+        """
+        charm_libs_path = self._project_dir / utils.get_lib_charm_path(charm_name)
+        if not charm_libs_path.is_dir():
+            return None
+        for api_version_path in charm_libs_path.iterdir():
+            lib_path = api_version_path / f"{lib_name}.py"
+            if lib_path.exists() and lib_path.is_file() or lib_path.is_symlink():
+                info = utils.get_lib_info(lib_path=lib_path)
+                if info.patch == -1:
+                    return None
+                return (info.api, info.patch)
+        return None
+
+    def write_lib(self, library: Library) -> None:
+        """Write the given library to disk.
+
+        :param library: A store library object with valid content.
+        """
+        if library.content is None:
+            # This should be considered an internal error.
+            raise ValueError("Library has no content.")
+        lib_path = self._project_dir / utils.get_lib_path(
```

Comment:
the library needs to be fetched, and stored in the lifecycle directories, not the project. If not the library will persist (which we have fetch-libs for).

Reply:
(Marking as resolved from external conversation. The spec is unclear, but it was determined that we do need to mutate the project directory here.)